### PR TITLE
Use C++ templates in `EvolveE` function

### PR DESCRIPTION
**Pull request by RemiLehe**
This is a follow-up of PR #616.

This updates the `EvolveE` function in the same way as `EvolveB`, using templates for the discretized stencil. In addition, the developer documentation is modified accordingly.

### CODE REVIEW

File: Source/FieldSolver/WarpXPushFieldsEM.cpp [hunk 1]
```diff
@@ -263,6 +271,7 @@ WarpX::EvolveE (int lev, PatchType patch_type, amrex::Real a_dt)
     const Real xmin = Geom(0).ProbLo(0);
 
     // Loop through the grids, and over the tiles within each grid
+/*
```
Comment (MaxThevenet) [inline Source/FieldSolver/WarpXPushFieldsEM.cpp]: Should we get rid of all this part?

Comment (RemiLehe) [reply Source/FieldSolver/WarpXPushFieldsEM.cpp]: Agreed, thanks! I just updated the PR.

File: Docs/source/developers/fields.rst [hunk 2]
```diff
@@ -69,11 +69,16 @@ The innermost step ``// Apply field solver on the FAB`` could be done with 3 nes
   amrex::ParallelFor(tex,
       [=] AMREX_GPU_DEVICE (int j, int k, int l)
       {
-          warpx_push_ex_yee(...);
+          Ex(i, j, k) += c2 * dt * (
+              - T_Algo::DownwardDz(By, coefs_z, n_coefs_z, i, j, k)
+              + T_Algo::DownwardDy(Bz, coefs_y, n_coefs_y, i, j, k)
+              - PhysConst::mu0 * jx(i, j, k) );
```
Comment (MaxThevenet) [inline Docs/source/developers/fields.rst]: :sparkling_heart: Thanks for updating this :).

**Comment by MaxThevenet** [comment]
Could you fix the conflicts? Then I'll merge.

**Pull request merged**